libdnf-cli/utils/utf8.cpp

```diff
@@ -0,0 +1,164 @@
+#include "utf8.hpp"
+
+#include <clocale>
+#include <cstring>
+#include <cwchar>
+
+
+namespace libdnf_cli::utils::utf8 {
+
+
+/// return length of an utf-8 encoded string
+std::size_t length(const std::string & str) {
+    std::size_t result = 0;
+
+    // pointers to the current position (defaults to begin) and the end of the input string
+    const char * ptr = str.c_str();
+    const char * end = ptr + strlen(ptr);
+
+    // multi-byte string state; required by mbrtowc()
+    std::mbstate_t state = std::mbstate_t();
+
+    // the wide char read from the input string
+    wchar_t wide_char = 0;
+
+    while (ptr < end) {
+        auto bytes = std::mbrtowc(&wide_char, ptr, MB_CUR_MAX, &state);
+        if (bytes <= 0) {
+            break;
+        }
+
+        // increase character count
+        result += 1;
+
+        // move the input string pointer by number of bytes read into the wide_char
+        ptr += bytes;
+    }
+
+    return result;
+}
+
+
+/// return printable width of an utf-8 encoded string (considers non-printable and wide characters)
+std::size_t width(const std::string & str) {
+    std::size_t result = 0;
+
+    // pointers to the current position (defaults to begin) and the end of the input string
+    const char * ptr = str.c_str();
+    const char * end = ptr + strlen(ptr);
```

Comment:
```
auto ptr = &str.front();
auto end = &str.back();

```
and change condition in while to <= .
The same on more places.

Reply:
fixed